packages/opentelemetry-resources/src/platform/node/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { CONTAINER_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+export const cache = {
+  readFileAsync: util.promisify(fs.readFile),
+};
+
+class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  readonly ECS_METADATA_KEY_V4 = 'ECS_CONTAINER_METADATA_URI_V4';
+  readonly ECS_METADATA_KEY_V3 = 'ECS_CONTAINER_METADATA_URI';
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let hostName;
+    try {
+      hostName = os.hostname();
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read host name: ${e.message}`
+      );
+    }
+
+    const containerId = await (await this._getContainerId(config)).toString();
```

Comment:
Here, there is another writing style:
```const [containerId] = await Promise.all([this._getContainerId(config)]);```
But I am not so sure which one performs better in typescript

Reply:
Are you getting a typescript error with just one await? There might be something wrong with your environment since await on a Promise returns the string, not a promise

Reply:
Yes, I got an error in that way when assigning containerId to resource constants, then I searched and found these 3 plausible ways.

Reply:
Oh, one thing I noticed that in _getContainerId method I used ```return``` rather than ```resolve```, I will modify and make another commit